### ISSUE TITLE
Add raven-testdata as submodule

### ISSUE THREAD

**Issue by cjauvin** [CSHS-CWRA/RavenPy#64]
Since the users who want to run the test suite have been using git anyway, I'm thinking that it would make sense (and make it easier for them) to connect `raven-testdata` as a git submodule, in order to solve the problem of having to "git clone it somewhere and set an env var to that somewhere".

**Comment by Zeitsperre** [comment]
If we were to do that, would this necessitate users having to clone the repo from GitHub in order to have the testdata/tests? If I were to install RavenPy from PyPI, we circumvent the `git clone` step, so I imagine it would only be for a development install via GitHub?

This would work, but I wonder if it might be better to use [`gitpython`](https://gitpython.readthedocs.io/en/stable/intro.html) to clone the repo instead. ROOCS is looking into doing just this for a similar problem.

**Comment by huard** [comment]
@cjauvin Do you have experience with submodule ? As mentionned by Trevor, another project we're collaborating with used to do that and just decided to switch because it proved troublesome.

**Comment by cjauvin** [comment]
The scenario I have in mind is this: the user git clones the code (because she wants to work on it for instance, or run the test suite) and she then wants to pip install it in editable mode in her current environment (although this is perhaps not necessary because the tests could run without the package being installed I guess? not sure) and while also taking the occasion to install the Raven binaries (which is now made possible in editable mode by my other [PR](https://github.com/CSHS-CWRA/RavenPy/pull/62)). So my idea here is that the testdata-as-submodule setup makes sense here, in this context where we're already using git.

In other words I'm targetting users like Richard, Julie and Mohammad, each of whom I spent a lot of time in the past weeks explaining things like: put the binaries in the path of the venv, export that variable in your shell, bla bla bla. Anything that I can imagine to simplify this process is a win in my mind!

**Comment by cjauvin** [comment]
The confusion might stem from the many different uses of pip that we make in this project:
1. A way to get a package from PyPI
2. A way to install some code I have on my disk (in editable mode or not)
3. A way to get some required binaries and put them at the right place

**Comment by tlvu** [comment]
I also agree with @huard and @Zeitsperre that git submodule is a feature of last resort.  If you want to know why I can elaborate further but there are many reasons so don't want to make this post too long.

I though we already have the `get_file` method used for the tutorial notebook and I vaguely remember you added a way to get all the files in bulk instead of getting files one by one.  Why don't we leverage that instead?

This is just for development I assume?  Then it does not even have to be integrated into RavenPy code.  A simple new Makefile target that just call git clone is good enough?

**Comment by Zeitsperre** [comment]
> I though we already have the `get_file` method used for the tutorial notebook and I vaguely remember you added a way to get all the files in bulk instead of getting files one by one. Why don't we leverage that instead?

The issue with it (as much as I really liked how simple it was), was it run up against the GitHub API call limits. If run too often, it will simply crash (with an error 403, I imagine). It's great for a handful of files (especially for running notebook tests), but not for development.

> This is just for development I assume? Then it does not even have to be integrated into RavenPy code. A simple new Makefile target that just call git clone is good enough?

I'm curious to see where the team at ROOCS goes with `gitpython`. I can find the implementation they've made and link to it here. `makefile` makes sense if we don't care to run tests off of a PyPI-sourced RavenPy, but being able to run the tests from a PyPI installation would be ideal, IMO.

**Comment by cjauvin** [comment]
The idea proposed here (submodule) is simply a way to make the already existing mechanism of "cloning the testdata repo + setting up an envvar to say where it is" a bit friendlier for the dev user: nothing more, nothing less.

**Comment by tlvu** [comment]
> The issue with it (as much as I really liked how simple it was), was it run up against the GitHub API call limits. If run too often, it will simply crash (with an error 403, I imagine). It's great for a handful of files (especially for running notebook tests), but not for development.

This is odd, why would we max out the API call limit?  I assume we do not repeatably redownload the same file on each `get_file` call?  We do have a local cache?  And for the `md5checksum` file that can assert if the local cache is up-to-date, we can check if the file is older than 1 or 2 hours before re-downloading it so we do not hit Github for this file on each `get_file` invocation?

> being able to run the tests from a PyPI installation would be ideal, IMO.

Oh right this would be cool indeed !  Maybe the `gitpython` approach will help us hook this into `setup.py`.

**Comment by Zeitsperre** [comment]
The team at ROOCS have already integrated it into their `roocs-utils` helper library here: https://github.com/roocs/roocs-utils/blob/ce3b76d3337ef08c29675ed36f1be638bfbe0c54/tests/conftest.py#L37

**Comment by tlvu** [comment]
> The team at ROOCS have already integrated it into their `roocs-utils` helper library here: https://github.com/roocs/roocs-utils/blob/ce3b76d3337ef08c29675ed36f1be638bfbe0c54/tests/conftest.py#L37

Oh geez, it they did not hardcode `MINI_ESGF_CACHE_DIR`, `TEST_DATA_REPO_URL` we could have just use their file as-is !

**Comment by Zeitsperre** [comment]
Oh yeah, that's true. We don't install `roocs-utils` by default (it's bundled with CLISOPS) so we might need to make our own? So many ways of fetching test data...
